### ISSUE TITLE
Allowing Python agents

### ISSUE THREAD

**Issue by pascalwhoop** [powertac/powertac-core#6]
Hi,
I forgot to introduce myself [earlier](https://github.com/powertac/server-distribution/pull/7), so I'll start with that first. I am a student with Prof. Ketter and I intend to write a thesis about RF agents this semester, evaluating it against the PowerTAC simulation. Whether it joins the competition remains to be discussed.

That out of the way, I would like to help by expanding the current platform to allow for non-java clients (especially Python in my case). The current JMS pipeline prohibits that AFAIK.

Therefore I'd like to enquire where I should start. I was thinking about 2 possible paths:

1. Using the current sample-agent and redesigning it to expose an API (web hooks, RPC, REST, that remains TBD)
2. Adding a second way of connecting to the server directly. I assume this would be the "cleanest" approach but it would maybe require a significant amount of reimplementing existing client code on the new language (and possibly having to do this repeatedly for further languages down the road)

Regarding the connection (be it via the java client as a middle-man or directly to the server) I would recommend [GRPC](grpc.io). It's nice because it will then also allow other teams to use other languages and it will generate the connective code for them. 

Best regards,
Pascal

**Comment by jecollins** [comment]
First, this issue should be moved to the powertac-server repo.

Second, this problem is exactly what the [broker-adapter](https://github.com/powertac/broker-adapter) is supposed to address. Is that approach inadequate?

**Comment by pascalwhoop** [comment]
1. Sorry, I didn't know that until after our call this week. I'll close this one. 

2. Yes and no. 

As I understand it, the adapter taps into the communication between server and client and relies on the external system to answer the server but still uses the existing code of the broker-core to handle all the meta-stuff. That's perfect and is exactly what I would like to avoid reimplementing. 

BUT, it relies on the local FS while I would like a python client to also be possibly remote, i.e. communicate via sockets. It allows me to run the entire "Java Block" somewhere else, e.g. the cloud. It also allows for having multiple simulations run in parallel by a single decision making agent. I expect the feed-forward time of the policy determining NN to be well under 5s --> tac stepping is the bottleneck and I can run several at once.

P.S. I could't find the `IpcAdapter` class or the `sendRawMessage` method anywhere. It's referenced in the `CharStreamAdapter` class. The IpcAdapter should be the one that hooks into the brokers MessageDispatcher  right?

**Comment by jecollins** [comment]
The IpcAdapter interface is on the RBroker branch of broker-core. There are several other changes in that branch, such as tracking which messages to process and which to pass through as raw xml, so it's not just the IpcAdapter that's missing. I'll make a pull request, but the merge/test will take a bit of effort.

**Comment by pascalwhoop** [comment]
I’m happy to help with that merging. Curious, searching for that in the PowerTAC org in github didn’t give any results. It seems search only works on the master branches,. Good to know

**Comment by jecollins** [comment]
I'm not worried about the merge, but about testing it. I'm still thinking
about that, and I really want to be home to do it. We'll be back on
Wednesday.

**Comment by pascalwhoop** [comment]
@jecollins I'm happy to help in any way then. I've actually also been thinking about a different approach and I'd like to hear what you think about it. Basically it would expose the `handleMessage` methods in the three classes of the sample-broker to the outside. This way, it would basically be like a normal Java Client, just that the methods get processed somewhere else. I'd just like to hear why you went with the direct XML passing instead of a RPC based approach.

**Comment by jecollins** [comment]
The reason I did not want to just expose the handleMessage() methods is that I did not want to burden developers using non-JVM languages with the need to pull apart JVM data structures. The raw-xml approach seems more language-agnostic. Keep in mind that this approach does not require the Java core to unmarshal and then re-marshal the xml - it just looks at the initial tag and decides whether to forward it as a string. So the unmarshaling work is delegated to the implementation, and most languages have xml libraries to take care of that.
